Update EDDN responder shipyard data to include `allowCobraMkIV` property

Tkael opened the pull request:
Ref. https://github.com/EDCD/EDDN/pull/98
Update the cAPI profile parser to parse the "capabilities" property, e.g.
```cs
  "capabilities": {
   "AllowCobraMkIV": false,
   "Horizons": false
  }
```

Resolves #1881.

Tkael committed:
Update EDDN responder shipyard data to include `allowCobraMkIV` property

Ref. https://github.com/EDCD/EDDN/pull/98
Update the cAPI profile parser to parse the "capabilities" property, e.g.
```cs
  "capabilities": {
   "AllowCobraMkIV": false,
   "Horizons": false
  }
```

Resolves #1881.

richardbuckle commented on CompanionAppService/CompanionAppService.cs:
```diff
@@ -647,6 +647,16 @@ public static Profile ProfileFromJson(JObject json, DateTime timestamp)
+                {
+                    var contexts = new ProfileContexts 
+                    { 
+                        allowCobraMkIV = (bool) json["commander"]["capabilities"]["AllowCobraMkIV"], 
```
No space after cast please

Tkael replied:
Fixed by [`6562958` (#1882)](https://github.com/EDCD/EDDI/pull/1882/commits/656295801cad96ed95067b4851a563ba61b4fd22).

richardbuckle commented on Events/MarketInformationUpdatedEvent.cs:
```diff
@@ -30,9 +32,10 @@ public class MarketInformationUpdatedEvent : Event
         /// <summary>The timestamp recorded for this event must be generated from game or server data.
         /// System time (e.g. DateTime.UtcNow) cannot be trusted for reporting to EDDN and may not be used.</summary>
-        public MarketInformationUpdatedEvent(DateTime timestamp, bool inHorizons, string starSystem, string stationName, long? marketId, List<CommodityMarketQuote> commodities, List<string> prohibitedCommodities, List<Module> outfitting, List<Ship> shipyard) : base(timestamp, NAME)
+        public MarketInformationUpdatedEvent(DateTime timestamp, bool inHorizons, bool? allowCobraMkIV, string starSystem, string stationName, long? marketId, List<CommodityMarketQuote> commodities, List<string> prohibitedCommodities, List<Module> outfitting, List<Ship> shipyard) : base(timestamp, NAME)
```
Yuck to an optional bool in the middle of a the non-optional args. Make it required bool and change all call sites.

Tkael replied:
Fixed by [`6562958` (#1882)](https://github.com/EDCD/EDDI/pull/1882/commits/656295801cad96ed95067b4851a563ba61b4fd22).
I kept it as an optional bool but moved it to the end of the .ctor and gave it a default value of null.
A null value is appropriate, in my opinion, for messages not involving the shipyard.

richardbuckle commented on CompanionAppService/CompanionAppService.cs:
```diff
@@ -647,6 +647,16 @@ public static Profile ProfileFromJson(JObject json, DateTime timestamp)
+                    var contexts = new ProfileContexts 
+                    { 
+                        allowCobraMkIV = (bool) json["commander"]["capabilities"]["AllowCobraMkIV"], 
+                        inHorizons = (bool) json["commander"]["capabilities"]["Horizons"] 
```
Ditto

Tkael replied:
Fixed by [`6562958` (#1882)](https://github.com/EDCD/EDDI/pull/1882/commits/656295801cad96ed95067b4851a563ba61b4fd22).

richardbuckle commented on CompanionAppService/Profile.cs:
```diff
@@ -29,5 +29,17 @@ public class Profile
+        public ProfileContexts contexts { get; set; }
+    }
+
+    public class ProfileContexts
```
Seems very heavyweight for just two flags.

richardbuckle reviewed (commented):
Hope my notes make sense.

richardbuckle reviewed (changes requested):
Per my notes

Tkael committed:
Per @richardbuckle's comments.
Removed spacing between casts and values.
Moved `inHorizons` and `allowCobraMkIV` context parameters to the end of the .ctor and defaulted bool? `allowCobraMkIV` to null (it is truly not applicable for outfitting and commodity quote messages so using null in those messages is appropriate)

GioVAX commented on CompanionAppService/CompanionAppService.cs:
```diff
@@ -647,6 +647,16 @@ public static Profile ProfileFromJson(JObject json, DateTime timestamp)
                 Profile.docked = (bool)json["commander"]["docked"];
                 Profile.alive = (bool)json["commander"]["alive"];
 
+                if (json["commander"]["capabilities"] != null)
```
I know this might not be the best time to discuss this, but why aren't we using something like Newtonsoft DeserializeObject instead of parsing each single property?

Tkael replied:
I'd be happy to go with that option too but that would require some more significant changes to our overall parsing approach than I wanted to implement at this time.

GioVAX replied:
Got it.

GioVAX commented on CompanionAppService/Profile.cs:
```diff
@@ -29,5 +29,17 @@ public class Profile
+        public ProfileContexts contexts { get; set; }
+    }
+
+    public class ProfileContexts
```
I am probably missing some context here: what's the benefit of a separate class instead of adding 2 more props to the Profile class?

Tkael replied:
Firstly, because this more closely follows the structure of the profile received from FDev. Secondly, because it makes sense to me to group context information together. Thirdly, because I expect that additional contexts will be added in later updates (the Odyssey expansion, for example, will likely add a new property here).

GioVAX replied:
Accepted for the second point.
As for the other 2 points, I believe we should have our own internal model, and map from the FDev model accordingly.
